### PR TITLE
Keep resident runner connected on iOS when screen locks in debug mode

### PR DESCRIPTION
**Pull request by xster**
For https://github.com/flutter/flutter/issues/9351

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 1]
```diff
@@ -4,8 +4,11 @@
 
 #include "flutter/shell/platform/darwin/ios/framework/Headers/FlutterAppDelegate.h"
 #include "flutter/shell/platform/darwin/ios/framework/Headers/FlutterViewController.h"
+#include "lib/ftl/logging.h"
 
-@implementation FlutterAppDelegate
+@implementation FlutterAppDelegate {
+  UIBackgroundTaskIdentifier debugBackgroundTask;
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: The ObjC convention would be `_debugBackgroundTask`.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 2]
```diff
@@ -26,4 +29,27 @@ - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
...
+- (void)applicationDidEnterBackground:(UIApplication *)application {
+#if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG
+  // The following keeps the Flutter session alive when the device screen locks
+  // in debug mode. It allows continued use of features like hot reload and 
+  // taking screenshots once the device unlocks again.
+  //
+  // Note the name is not an identifier and multiple instances can exist. 
+  debugBackgroundTask = [application beginBackgroundTaskWithName:@"Flutter debug task"
+                                               expirationHandler:^{
+      FTL_DLOG(WARNING) << "\nThe OS has terminated the Flutter debug connection for being "
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: You need `FTL_LOG`. The `DLOG` variant will only be available in unoptimized builds. We release optimized builds.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 3]
```diff
@@ -26,4 +29,27 @@ - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
   }
 }
 
+- (void)applicationDidEnterBackground:(UIApplication *)application {
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: Nit: You can get rid of the overrides altogether in profile and release modes.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 4]
```diff
@@ -29,27 +29,27 @@ - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
...
+                          "inactive in the background for too long.\n\n"
+                          "There are no errors with your Flutter application.\n\n"
+                          "To reconnect, launch your application again via 'flutter run";
       }];
-#endif
 }
+#endif
 
-- (void)applicationWillEnterForeground:(UIApplication *)application {
 #if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: Can probably kill off the `#endif and this line`.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 5]
```diff
@@ -29,27 +29,27 @@ - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
...
-  [application endBackgroundTask: debugBackgroundTask];
-#endif
+- (void)applicationWillEnterForeground:(UIApplication *)application {
+  [application endBackgroundTask: _debugBackgroundTask];
 }
+#endif
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: `#endif  // FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG`

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm (hunk 4 above)
Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: ah ya you're right

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm (hunk 5 above)
Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 6]
```diff
@@ -26,4 +29,25 @@ - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
...
+                          "inactive in the background for too long.\n\n"
+                          "There are no errors with your Flutter application.\n\n"
+                          "To reconnect, launch your application again via 'flutter run";
+      }];
+}
+
+- (void)applicationWillEnterForeground:(UIApplication *)application {
+  [application endBackgroundTask: _debugBackgroundTask];
+}
+#endif // FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: cargo-culting nit: two spaces after the `#endif` elsewhere in the engine.

Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: Done 
(we need a formatter)

**Comment by cbracken** [comment]
lgtm

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm (hunk 6 above)
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: > (we need a formatter)

We use `clang-format` in `buildtools`. There are `clang-format` manifests in the the build root as well as engine repositories. You'd just have to specify the file config. I was just assuming you were using the formatter already.

**Pull request merged**